File: js/src/csv.ts
import * as fs from "fs";
import Papa from "papaparse";

export class Csv {
  static readonly SEPARATOR = ";";
  static readonly QUOTE_CHAR = '"';

  static readonly CSV_OK = 0;
  static readonly CSV_FILE_NOT_READABLE = 2;
  static readonly CSV_FILE_NOT_WRITABLE = 3;
  static readonly CSV_ROW_SIZE_MISMATCH = 4;
  static readonly CSV_COLUMN_SIZE_MISMATCH = 5;
  static readonly CSV_DIFFERENT_KEYS = 6;
  static readonly CSV_COLUMN_NOT_FOUND = 7;
  static readonly CSV_ROW_NOT_FOUND = 8;

  private _header: string[] = [];
  private _data: Record<string, string>[] = [];

  constructor(data?: string[][] | Record<string, string>[]) {
    if (!data || data.length === 0) {
      return;
    }

    const first = data[0];
    if (Array.isArray(first)) {
      const rows = data as string[][];
      this._header = rows[0].map(String);
      for (let i = 1; i < rows.length; i++) {
        const row = rows[i];
        if (row.length !== this._header.length) {
          throw new Error("All rows must have the same length as the header");
        }
        const obj: Record<string, string> = {};
        for (let j = 0; j < this._header.length; j++) {
          obj[this._header[j]] = String(row[j]);
        }
        this._data.push(obj);
      }
    } else {
      const rows = data as Record<string, string>[];
      const keys = Object.keys(first);
      for (const row of rows) {
        const rowKeys = Object.keys(row);
        if (rowKeys.length !== keys.length || !rowKeys.every((k, i) => k === keys[i])) {
          throw new Error("All rows must have the same keys in the same order");
        }
      }
      this._header = keys.map(String);
      this._data = rows.map((row) => {
        const obj: Record<string, string> = {};
        for (const k of keys) {
          obj[k] = String(row[k]);
        }
        return obj;
      });
    }
  }

  getHeader(): string[] {
    return [...this._header];
  }

  getData(): Record<string, string>[] {
    return this._data.map((row) => ({ ...row }));
  }

  totalRows(): number {
    return this._data.length;
  }

  totalColumns(): number {
    return this._header.length;
  }

  getRow(index: number): Record<string, string> {
    return { ...this._data[index] };
  }

  getColumn(columnName: string): string[] {
    return this._data.map((row) => row[columnName]);
  }

  getCell(columnName: string, index: number): string {
    return this._data[index][columnName];
  }

  addRow(row: string[] | Record<string, string>): number {
    if (Array.isArray(row)) {
      if (row.length !== this._header.length) {
        return Csv.CSV_ROW_SIZE_MISMATCH;
      }
      const obj: Record<string, string> = {};
      for (let i = 0; i < this._header.length; i++) {
        obj[this._header[i]] = String(row[i]);
      }
      this._data.push(obj);
      return Csv.CSV_OK;
    } else {
      if (Object.keys(row).length !== this._header.length) {
        return Csv.CSV_ROW_SIZE_MISMATCH;
      }
      const rowKeys = Object.keys(row);
      if (rowKeys.length !== this._header.length || !rowKeys.every((k, i) => k === this._header[i])) {
        return Csv.CSV_DIFFERENT_KEYS;
      }
      this._data.push({ ...row });
      return Csv.CSV_OK;
    }
  }

  addColumn(columnName: string, column: string[]): number {
    if (column.length !== this._data.length) {
      return Csv.CSV_COLUMN_SIZE_MISMATCH;
    }
    this._header.push(columnName);
    for (let i = 0; i < this._data.length; i++) {
      this._data[i][columnName] = String(column[i]);
    }
    return Csv.CSV_OK;
  }

  setCellValue(columnName: string, index: number, value: string): void {
    this._data[index][columnName] = value;
  }

  removeRow(index: number): number {
    if (index < 0 || index >= this._data.length) {
      return Csv.CSV_ROW_NOT_FOUND;
    }
    this._data.splice(index, 1);
    return Csv.CSV_OK;
  }

  removeColumn(columnName: string): number {
    const idx = this._header.indexOf(columnName);
    if (idx === -1) {
      return Csv.CSV_COLUMN_NOT_FOUND;
    }
    this._header.splice(idx, 1);
    for (const row of this._data) {
      delete row[columnName];
    }
    return Csv.CSV_OK;
  }

  loadFromString(
    csvString: string,
    hasHeader = true,
    separator = Csv.SEPARATOR,
    quoteChar = Csv.QUOTE_CHAR
  ): number {
    this._header = [];
    this._data = [];

    const trimmed = csvString.trim();
    if (!trimmed) {
      return Csv.CSV_OK;
    }

    const result = Papa.parse<string[]>(trimmed, {
      delimiter: separator,
      quoteChar: quoteChar,
      skipEmptyLines: true,
    });

    const rows = result.data;
    if (rows.length === 0) {
      return Csv.CSV_OK;
    }

    if (hasHeader) {
      this._header = rows[0].map(String);
      for (let i = 1; i < rows.length; i++) {
        const obj: Record<string, string> = {};
        for (let j = 0; j < this._header.length; j++) {
          obj[this._header[j]] = String(rows[i][j] ?? "");
        }
        this._data.push(obj);
      }
    } else {
      this._header = rows[0].map((_, idx) => String(idx));
      for (const row of rows) {
        const obj: Record<string, string> = {};
        for (let j = 0; j < this._header.length; j++) {
          obj[this._header[j]] = String(row[j] ?? "");
        }
        this._data.push(obj);
      }
    }

    return Csv.CSV_OK;
  }

  toString(
    hasHeader = true,
    separator = Csv.SEPARATOR,
    quoteChar = Csv.QUOTE_CHAR
  ): string {
    const rows: string[][] = [];

    if (hasHeader) {
      rows.push(this._header);
    }

    for (const row of this._data) {
      rows.push(this._header.map((k) => row[k] ?? ""));
    }

    return Papa.unparse(rows, {
      delimiter: separator,
      quoteChar: quoteChar,
      newline: "\n",
    });
  }

  load(
    filePath: string,
    hasHeader = true,
    separator = Csv.SEPARATOR,
    quoteChar = Csv.QUOTE_CHAR
  ): number {
    let content: string;
    try {
      content = fs.readFileSync(filePath, "utf-8");
    } catch {
      return Csv.CSV_FILE_NOT_READABLE;
    }
    return this.loadFromString(content, hasHeader, separator, quoteChar);
  }

  save(
    filePath: string,
    hasHeader = true,
    separator = Csv.SEPARATOR,
    quoteChar = Csv.QUOTE_CHAR
  ): number {
    try {
      const content = this.toString(hasHeader, separator, quoteChar);
      fs.writeFileSync(filePath, content, "utf-8");
      return Csv.CSV_OK;
    } catch {
      return Csv.CSV_FILE_NOT_WRITABLE;
    }
  }
}

File: cpp/src/csv.cpp
#include "csv.h"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace mc {

namespace {

// Full-string CSV parser: handles quoted fields (including multi-line quoted
// fields), doubled-quote escaping, CRLF, and custom delimiter / quote char.
std::vector<std::vector<std::string>> ParseCsv(const std::string& content,
                                               char delim, char qc) {
  std::vector<std::vector<std::string>> rows;
  std::vector<std::string> row;
  std::string field;
  bool in_quotes = false;
  const size_t len = content.size();

  for (size_t i = 0; i <= len; ++i) {
    // Flush on end-of-input
    if (i == len) {
      row.push_back(std::move(field));
      // Discard a trailing empty row produced by a trailing newline
      if (!(row.size() == 1 && row[0].empty())) {
        rows.push_back(std::move(row));
      }
      break;
    }

    char c = content[i];

    if (in_quotes) {
      if (c == qc) {
        if (i + 1 < len && content[i + 1] == qc) {
          // Doubled quote → literal quote character
          field += qc;
          ++i;
        } else {
          in_quotes = false;
        }
      } else {
        field += c;
      }
    } else {
      if (c == qc) {
        in_quotes = true;
      } else if (c == delim) {
        row.push_back(std::move(field));
        field.clear();
      } else if (c == '\n') {
        row.push_back(std::move(field));
        field.clear();
        if (!row.empty()) {
          rows.push_back(std::move(row));
          row.clear();
        }
      } else if (c == '\r') {
        // Skip CR; '\n' will flush the row
      } else {
        field += c;
      }
    }
  }

  return rows;
}

// Serialise a single field, quoting it when it contains special characters.
std::string SerialiseField(const std::string& f, char delim, char qc) {
  bool needs_quote = false;
  for (char c : f) {
    if (c == delim || c == qc || c == '\n' || c == '\r') {
      needs_quote = true;
      break;
    }
  }
  if (!needs_quote) return f;

  std::string out;
  out.reserve(f.size() + 2);
  out += qc;
  for (char c : f) {
    if (c == qc) out += qc;  // double the quote
    out += c;
  }
  out += qc;
  return out;
}

}  // namespace

// ── Constructor ───────────────────────────────────────────────────────────

Csv::Csv(std::vector<std::vector<std::string>> data) {
  if (data.empty()) return;
  header_ = std::move(data[0]);
  rows_.reserve(data.size() - 1);
  for (size_t i = 1; i < data.size(); ++i) {
    if (data[i].size() != header_.size()) {
      throw std::invalid_argument(
          "All rows must have the same length as the header");
    }
    rows_.push_back(std::move(data[i]));
  }
}

// ── Private helpers ───────────────────────────────────────────────────────

int Csv::ColumnIndex(const std::string& name) const {
  auto it = std::find(header_.begin(), header_.end(), name);
  if (it == header_.end()) return -1;
  return static_cast<int>(it - header_.begin());
}

// ── Accessors ─────────────────────────────────────────────────────────────

std::vector<std::string> Csv::GetHeader() const { return header_; }

std::vector<std::map<std::string, std::string>> Csv::GetData() const {
  std::vector<std::map<std::string, std::string>> result;
  result.reserve(rows_.size());
  for (const auto& row : rows_) {
    std::map<std::string, std::string> m;
    for (size_t i = 0; i < header_.size(); ++i) m[header_[i]] = row[i];
    result.push_back(std::move(m));
  }
  return result;
}

int Csv::TotalRows() const { return static_cast<int>(rows_.size()); }
int Csv::TotalColumns() const { return static_cast<int>(header_.size()); }

std::map<std::string, std::string> Csv::GetRow(int index) const {
  if (index < 0 || index >= TotalRows())
    throw std::out_of_range("Row index out of range");
  std::map<std::string, std::string> m;
  const auto& row = rows_[static_cast<size_t>(index)];
  for (size_t i = 0; i < header_.size(); ++i) m[header_[i]] = row[i];
  return m;
}

std::vector<std::string> Csv::GetColumn(const std::string& column_name) const {
  int idx = ColumnIndex(column_name);
  if (idx < 0) throw std::invalid_argument("Column not found: " + column_name);
  std::vector<std::string> result;
  result.reserve(rows_.size());
  for (const auto& row : rows_)
    result.push_back(row[static_cast<size_t>(idx)]);
  return result;
}

std::string Csv::GetCell(const std::string& column_name, int index) const {
  int col = ColumnIndex(column_name);
  if (col < 0) throw std::invalid_argument("Column not found: " + column_name);
  if (index < 0 || index >= TotalRows())
    throw std::out_of_range("Row index out of range");
  return rows_[static_cast<size_t>(index)][static_cast<size_t>(col)];
}

// ── Mutators ──────────────────────────────────────────────────────────────

int Csv::AddRow(const std::vector<std::string>& row) {
  if (static_cast<int>(row.size()) != TotalColumns()) return kRowSizeMismatch;
  rows_.push_back(row);
  return kOk;
}

int Csv::AddRow(const std::map<std::string, std::string>& row) {
  if (static_cast<int>(row.size()) != TotalColumns()) return kRowSizeMismatch;
  for (const auto& key : header_) {
    if (row.find(key) == row.end()) return kDifferentKeys;
  }
  std::vector<std::string> new_row;
  new_row.reserve(header_.size());
  for (const auto& key : header_) new_row.push_back(row.at(key));
  rows_.push_back(std::move(new_row));
  return kOk;
}

int Csv::AddColumn(const std::string& column_name,
                   const std::vector<std::string>& column) {
  if (static_cast<int>(column.size()) != TotalRows()) return kColumnSizeMismatch;
  header_.push_back(column_name);
  for (size_t i = 0; i < rows_.size(); ++i) rows_[i].push_back(column[i]);
  return kOk;
}

void Csv::SetCellValue(const std::string& column_name, int index,
                       const std::string& value) {
  int col = ColumnIndex(column_name);
  if (col < 0) return;
  if (index < 0 || index >= TotalRows()) return;
  rows_[static_cast<size_t>(index)][static_cast<size_t>(col)] = value;
}

int Csv::RemoveRow(int index) {
  if (index < 0 || index >= TotalRows()) return kRowNotFound;
  rows_.erase(rows_.begin() + index);
  return kOk;
}

int Csv::RemoveColumn(const std::string& column_name) {
  int col = ColumnIndex(column_name);
  if (col < 0) return kColumnNotFound;
  header_.erase(header_.begin() + col);
  for (auto& row : rows_)
    row.erase(row.begin() + col);
  return kOk;
}

// ── Serialisation ─────────────────────────────────────────────────────────

int Csv::LoadFromString(const std::string& csv_string, bool has_header,
                        char separator, char quote_char) {
  header_.clear();
  rows_.clear();

  // Strip UTF-8 BOM if present
  std::string content = csv_string;
  if (content.size() >= 3 &&
      static_cast<unsigned char>(content[0]) == 0xEF &&
      static_cast<unsigned char>(content[1]) == 0xBB &&
      static_cast<unsigned char>(content[2]) == 0xBF) {
    content = content.substr(3);
  }

  auto parsed = ParseCsv(content, separator, quote_char);
  if (parsed.empty()) return kOk;

  if (has_header) {
    header_ = std::move(parsed[0]);
    rows_.reserve(parsed.size() - 1);
    for (size_t i = 1; i < parsed.size(); ++i) {
      parsed[i].resize(header_.size());  // pad/truncate to match header
      rows_.push_back(std::move(parsed[i]));
    }
  } else {
    size_t width = parsed[0].size();
    header_.reserve(width);
    for (size_t i = 0; i < width; ++i) header_.push_back(std::to_string(i));
    rows_.reserve(parsed.size());
    for (auto& r : parsed) {
      r.resize(width);
      rows_.push_back(std::move(r));
    }
  }

  return kOk;
}

std::string Csv::ToString(bool has_header, char separator,
                          char quote_char) const {
  std::ostringstream oss;

  if (has_header) {
    for (size_t i = 0; i < header_.size(); ++i) {
      if (i > 0) oss << separator;
      oss << SerialiseField(header_[i], separator, quote_char);
    }
    oss << '\n';
  }

  for (const auto& row : rows_) {
    for (size_t i = 0; i < row.size(); ++i) {
      if (i > 0) oss << separator;
      oss << SerialiseField(row[i], separator, quote_char);
    }
    oss << '\n';
  }

  return oss.str();
}

int Csv::Load(const std::string& file_name, bool has_header, char separator,
              char quote_char) {
  std::ifstream f(file_name);
  if (!f.is_open()) return kFileNotReadable;
  std::string content((std::istreambuf_iterator<char>(f)),
                      std::istreambuf_iterator<char>());
  return LoadFromString(content, has_header, separator, quote_char);
}

int Csv::Save(const std::string& file_name, bool has_header, char separator,
              char quote_char) const {
  std::ofstream f(file_name);
  if (!f.is_open()) return kFileNotWritable;
  f << ToString(has_header, separator, quote_char);
  return f.fail() ? kFileNotWritable : kOk;
}

}  // namespace mc
